data/provider/src/main/java/tech/pegasys/teku/api/ChainDataProvider.java

```diff
@@ -173,26 +172,17 @@ public SafeFuture<Optional<StateAndMetaData>> getBeaconStateAndMetadata(
     return defaultStateSelectorFactory.defaultStateSelector(stateIdParam).getState();
   }
 
-  public SafeFuture<GetAllBlocksAtSlotResponse> getAllBlocksAtSlot(final String slot) {
-    if (slot.startsWith("0x")) {
-      throw new BadRequestException(
-          String.format("block roots are not currently supported: %s", slot));
-    } else {
-      return defaultBlockSelectorFactory
-          .nonCanonicalBlocksSelector(UInt64.valueOf(slot))
-          .getBlocks()
-          .thenApply(
-              blockList -> {
-                final Set<SignedBeaconBlockWithRoot> blocks =
-                    blockList.stream()
-                        .map(ObjectAndMetaData::getData)
-                        .map(SignedBeaconBlockWithRoot::new)
-                        .collect(Collectors.toSet());
-
-                return new GetAllBlocksAtSlotResponse(
-                    getVersionAtSlot(UInt64.valueOf(slot)), blocks);
-              });
-    }
+  public SafeFuture<AllBlocksAtSlotData> getAllBlocksAtSlot(final UInt64 slot) {
+    return defaultBlockSelectorFactory
+        .nonCanonicalBlocksSelector(slot)
+        .getBlocks()
+        .thenApply(
+            blockList -> {
+              final List<SignedBeaconBlock> blocks =
+                  blockList.stream().map(ObjectAndMetaData::getData).collect(Collectors.toList());
+
+              return new AllBlocksAtSlotData(spec.atSlot(slot).getMilestone(), blocks);
```

Comment:
if you were to pass in the blockList, which is ObjectAndMetadata, into the constructor, you'd have milestone and could create response straight from that...

Reply:
Do you mean pass `blockList` into `AllBlocksAtSlotData` constructor?

Reply:
yep

Reply:
That would make testing more painful, but could do? Is there much benefit, since I still need to pass in the milestone as it is now?

Reply:
milestone is in the metadata, can just have 2 constructors...

Reply:
It is a `List<BlockAndMetaData>`, should it be assumed that all in the list have the same milestone, so I just take any?